### PR TITLE
Add support for FIRST, LAST Aggregate Functions

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#6444](https://github.com/apache/arrow-datafusion/issues/6444).

# Rationale for this change

# What changes are included in this PR?
This PR adds support for `FIRST_VALUE`, `LAST_VALUE` ordering sensitive aggregation functions. With the changes in this PR we now can run query below
```sql
SELECT a, b, FIRST_VALUE(c ORDER BY a DESC) as first_c
  FROM annotated_data_infinite2
```
This query will give the first value for the `Column c` when table is ordered by `a DESC`.

# Are these changes tested?
Yes, new tests are added to the `groupby.slt` file for new aggregate functions.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mustafasrepo** [comment]
SELF TODO: FIRST, LAST agg’de order by kullanilmiyorsa (veya onemsize) ignore et, window querysi ise verilen orderingi kabul etme. (Standarta bak beklenen davranis bu mu?)

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 1]
```diff
@@ -388,14 +389,21 @@ fn get_finest_requirement<
     Ok(result)
 }
 
+// Check whether aggregate function is ordering sensitive
+fn is_ordering_sensitive(aggr_expr: &Arc<dyn AggregateExpr>) -> bool {
```
Comment (berkaysynnada) [inline datafusion/core/src/physical_plan/aggregates/mod.rs:393]: Can you add more detailed comments here?

**Review by ozankabak — APPROVED** [review]
LGTM except one question/possible issue: Why does `FirstValueAccumulator`/`LastValueAccumulator` need a `count` attribute? The `ScalarValue` they store is initialized with an empty (null) value, isn't it? If the group never got an item by the time `evaluate` is called, it would be returning this null value, which seems like the desired behavior.

Maybe I am missing something and the `count` field is indeed necessary, but please explain to me why in our next meeting -- if it is actually necessary.

**Comment by mustafasrepo** [comment]
Merged upstream

**Pull request closed without merging**